### PR TITLE
Oidc authentication

### PR DESCRIPTION
**Pull request by remk**
The goal of this PR, is to allow testing of openpaas using oidc authentication.

For the time being it only replace the authentication. When we will have a running platform we update the scenarii.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,7 +40,10 @@ Environment variables:
  - `USER_COUNT` which is set to `1` by default
  - `HUMAN_ACTION_MIN_DELAY` which is set to `7` by default. The minimum amount of seconds between 2 human actions.
  - `HUMAN_ACTION_MAX_DELAY` which is set to `15` by default. The maximum amount of seconds between 2 human actions.
- - `AUTHENTICATION_STRATEGY` which is set to `basic` by default (you can use `lemonldap` instead)
+ - `AUTHENTICATION_STRATEGY` which is set to `basic` by default (you can use `lemonldap`, `oidc` or `pkce` instead)
+ - `OIDC_CLIENT` oidc client to use for `oidc` and `pkce` authentication strategies
+ - `OIDC_CALLBACK` url to redirect to after a successful login with `oidc` and `pkce` authentication strategies
+ - `PKCE_CODE_CHALLENGE_METHOD` which is set to `S256` by default. Which should not be changed at the moment.
```
Comment (rouazana) [inline README.md:46]: can we remove this? given the spec I think it's useless to make it configurable, and we do not handle it in the code to generate the hash

File: src/main/scala/com/linagora/openpaas/gatling/core/LoginSteps.scala [hunk 2]
```diff
@@ -1,29 +1,59 @@
...
+      .exec(OIDCSteps.login)
+      .exec(OIDCSteps.goToOpenPaaSApplication)
+
+    case AuthenticationStrategy.PKCE =>
+      exec(session => {
+        val codeChallengeVerifier = UUID.randomUUID().toString + UUID.randomUUID().toString + UUID.randomUUID().toString
+        val codeChallenge = Base64.getUrlEncoder.encodeToString(MessageDigest.getInstance("SHA-256").digest(codeChallengeVerifier.getBytes(Charsets.US_ASCII)))
+
+        session.set("oidc_state", UUID.randomUUID().toString + UUID.randomUUID().toString)
+          .set("pkce_code_challenge",	codeChallenge.replace("=", ""))
```
Comment (rouazana) [inline src/main/scala/com/linagora/openpaas/gatling/core/LoginSteps.scala:46]: can you put the replace upper where codeChallenge is defined? It's part of the spec https://tools.ietf.org/html/rfc7636#page-17

File: src/main/scala/com/linagora/openpaas/gatling/core/authentication/pkce/PKCESteps.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  private def extractAuthorizationCodeFromLocation(locationUrl: String): String = {
+    Uri.create(locationUrl.replace("/#/","/"))
+      .getEncodedQueryParams.asScala.find(_.getName == "code").get.getValue
+  }
+
+  def goToOpenPaaSApplication: HttpRequestBuilder =
+    http("Go to OpenPaaS application")
+      .get("/")
+      //.disableFollowRedirect
```
Comment (rouazana) [inline src/main/scala/com/linagora/openpaas/gatling/core/authentication/pkce/PKCESteps.scala:83]: remove?

**Review by rouazana — APPROVED** [review]

**Review by rouazana — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (rouazana) [inline README.md:46]: you miss to remove this line

File: src/test/scala/com/linagora/openpaas/gatling/OpenPaaSMixSimulation.scala [hunk 4]
```diff
@@ -32,12 +32,14 @@ class OpenPaaSMixSimulation extends Simulation {
...
     .during(ScenarioDuration) {
       randomSwitch(
-        25.0 -> exec(SendMessageScenari.generate()),
-        25.0 -> exec(SearchEventsScenari.generate()),
-        25.0 -> exec(SendEmailScenari.generate()),
-        25.0 -> exec(OpenContactScenari.generate())
+        //25.0 -> exec(SendMessageScenari.generate()),
+        33.3 -> exec(SearchEventsScenari.generate()),
+        33.3 -> exec(SendEmailScenari.generate()),
+        33.3 -> exec(OpenContactScenari.generate())
```
Comment (chibenwa) [inline src/test/scala/com/linagora/openpaas/gatling/OpenPaaSMixSimulation.scala:42]: I think I do spend more time on email than on other stuff to be honnest

File: src/main/scala/com/linagora/openpaas/gatling/unifiedinbox/scenari/SendEmailScenari.scala [hunk 5]
```diff
@@ -20,11 +20,11 @@ object SendEmailScenari {
...
-      .exec(generateJwtToken)
-      .exec(generateJwtToken)
-      .exec(generateJwtToken)
-      .exec(generateJwtToken)
-      .exec(generateJwtToken)
+      .exec(generateJwtTokenWithAuth)
+      .exec(generateJwtTokenWithAuth)
+      .exec(generateJwtTokenWithAuth)
+      .exec(generateJwtTokenWithAuth)
+      .exec(generateJwtTokenWithAuth)
```
Comment (chibenwa) [inline src/main/scala/com/linagora/openpaas/gatling/unifiedinbox/scenari/SendEmailScenari.scala:27]: Why getting the JWT token 5 times?

Why getting mailboxes twice below?

File: src/test/scala/com/linagora/openpaas/gatling/OpenPaaSMixSimulation.scala (hunk 4 above)
Comment (chibenwa) [reply src/test/scala/com/linagora/openpaas/gatling/OpenPaaSMixSimulation.scala:42]: I do other stuff in the webmail than just sending emails.

Any plan to add more actions?

**Review by huy-ta — COMMENTED** [review]
You need to rebase your commits before we can approve this.

File: src/main/scala/com/linagora/openpaas/gatling/unifiedinbox/scenari/SendEmailScenari.scala (hunk 5 above)
Comment (remk) [reply src/main/scala/com/linagora/openpaas/gatling/unifiedinbox/scenari/SendEmailScenari.scala:27]: It was a capture of what openpaas was doing.
We should update it, maybe the thing have improved a little bit.
But it is not in the scope of this ticket

**Comment by remk** [comment]
> You need to rebase your commits before we can approve this.

I let the fixup to make it easier for reviewers who  follow the PR.

**Comment by rouazana** [comment]
we will take eventual missing comments in other PR

**Pull request merged**